pkg/manifestcomparators/comp_no_new_required_fields.go

```diff
@@ -0,0 +1,142 @@
+)
+
+type noNewRequiredFields struct{}
+
+func NoNewRequiredFields() CRDComparator {
+	return noNewRequiredFields{}
+}
+
+func (noNewRequiredFields) Name() string {
+	return "NoNewRequiredFields"
+}
+
+func (noNewRequiredFields) WhyItMatters() string {
+	return "If new fields are required, then old clients will not function properly.  Even if CRD defaulting is used, " +
+		"CRD defaulting requires allowing an object with an empty or missing value to then get defaulted."
+}
+
+func (b noNewRequiredFields) Compare(existingCRD, newCRD *apiextensionsv1.CustomResourceDefinition) (ComparisonResults, error) {
+	if existingCRD == nil {
+		return ComparisonResults{
+			Name:         b.Name(),
+			WhyItMatters: b.WhyItMatters(),
+
+			Errors:   nil,
+			Warnings: nil,
+			Infos:    nil,
+		}, nil
+	}
+	errsToReport := []string{}
+
+	for _, newVersion := range newCRD.Spec.Versions {
+
+		existingVersion := GetVersionByName(existingCRD, newVersion.Name)
+		if existingVersion == nil {
+			continue
+		}
+
+		existingRequiredFields := map[string]sets.String{}
+		existingToSimpleLocation := map[*apiextensionsv1.JSONSchemaProps]*field.Path{}
+		SchemaHas(existingVersion.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+			func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, _ []*apiextensionsv1.JSONSchemaProps) bool {
+				existingRequiredFields[simpleLocation.String()] = sets.NewString(s.Required...)
+				existingToSimpleLocation[s] = simpleLocation
+				return false
+			})
+
+		// New fields can be required if they are wrapped inside new structs that are themselves optional.
+		// For instance, you cannot add .spec.thingy as required, but if you add .spec.top as optional and at the same
+		// time add .spec.top.thingy as required, this is allowed.
+		// Similar logic exists for adding an array with minlength > 0
+		newRequiredFields := sets.NewString()
+		newSimpleLocationToRequiredFields := map[string]sets.String{}
+		newToSimpleLocation := map[*apiextensionsv1.JSONSchemaProps]*field.Path{}
+		SchemaHas(newVersion.Schema.OpenAPIV3Schema, field.NewPath("^"), field.NewPath("^"), nil,
+			func(s *apiextensionsv1.JSONSchemaProps, fldPath, simpleLocation *field.Path, ancestors []*apiextensionsv1.JSONSchemaProps) bool {
+				newSimpleLocationToRequiredFields[simpleLocation.String()] = sets.NewString(s.Required...)
+				newToSimpleLocation[s] = simpleLocation
+
+				if s.Type == "array" {
+					// if it's an array, we have a different property to check.  A new array cannot be required unless it's ancestor is new.
+					if s.MinLength == nil || *s.MinLength == 0 {
+						// if there is no required length, this is fine
+						return false
+					}
+					// this means we're an array with a minLength, check to see if any parent wrapper is both new and optional.
+					var prevAncestor *apiextensionsv1.JSONSchemaProps
+					for i := len(ancestors) - 1; i >= 0; i-- {
+						ancestor := ancestors[i]
+						if !ancestor.Nullable {
```

Comment:
Where does this property come from? Do you have to set a tag on the type to have it come up as nullable? Which types would get this set automatically?
```
// +nullable
```
I'm not seeing the nullable property in the test cases anywhere so I think there's some nuance I've missed here?

Reply:
Adding the `// +nullable` give it to us.  See https://github.com/openshift/api/pull/1683

We should probably use it for pointers.  See https://swagger.io/docs/specification/data-models/data-types/ for the type definition.